cranelift/codegen/src/legalizer/heap.rs

```diff
@@ -180,45 +182,59 @@ fn static_addr(
         heap,
         addr_ty,
         offset,
-        offset_ty,
         pos.func,
         spectre_oob_comparison,
     );
 }
 
+fn cast_offset_to_pointer_ty(
+    offset: ir::Value,
+    offset_ty: ir::Type,
+    addr_ty: ir::Type,
+    pos: &mut FuncCursor,
+) -> ir::Value {
+    if offset_ty == addr_ty {
+        return offset;
+    }
+    // Note that using 64-bit heaps on a 32-bit host is not currently supported,
+    // would require at least a bounds check here to ensure that the truncation
+    // from 64-to-32 bits doesn't lose any upper bits. For now though we're
+    // mostly interested in the 32-bit-heaps-on-64-bit-hosts cast.
+    assert!(offset_ty.bits() < addr_ty.bits());
+
+    // Convert `offset` to `addr_ty`.
+    let extended_offset = pos.ins().uextend(addr_ty, offset);
+    let loc = pos.srcloc();
+    if let Some(values_labels) = pos.func.dfg.values_labels.as_mut() {
```

Comment:
Not strictly in-scope (sorry!) and I see you've just moved this code from below, but since we're looking at it now: we should have a method on the DFG for this -- maybe something like `add_value_label_alias()`? If not, at least a comment to the effect of "Add debug value-label alias so that debuginfo can name the extended value as the address".